File: 2021/src/day14/day14.js
const input = (raw) => {
  const [template, ruleList] = raw.trim().split("\n\n");

  const rules = new Map(ruleList.split("\n").map((str) => str.split(" -> ")));
  return { template, rules };
};

const getIncrementer = (map) => (key, count) => {
  map.set(key, (map.get(key) ?? 0) + count);
};

const doInsertions = (pairs, rules) => {
  const newPairs = new Map(pairs);

  const inc = getIncrementer(newPairs);

  for (const [pair, count] of pairs) {
    const [a, b] = pair.split("");
    const x = rules.get(pair);

    inc(`${a}${x}`, count);
    inc(`${x}${b}`, count);
    inc(pair, -count);
  }

  return newPairs;
};

const run = (raw, iterations) => {
  const { template, rules } = input(raw);

  let pairs = new Map();
  const inc = getIncrementer(pairs);
  for (let i = 0; i < template.length - 1; i += 1) {
    inc(`${template[i]}${template[i + 1]}`, 1);
  }

  for (let i = 0; i < iterations; i += 1) {
    pairs = doInsertions(pairs, rules);
  }

  const occurances = new Map();
  const oinc = getIncrementer(occurances);
  for (const [pair, count] of pairs) {
    oinc(pair[0], count);
  }
  oinc(template[template.length - 1], 1);

  const max = Math.max(...occurances.values());
  const min = Math.min(...occurances.values());

  return max - min;
};

export const part1 = (raw) => {
  return run(raw, 10);
};
export const part2 = (raw) => {
  return run(raw, 40);
};

File: 2021/src/day14/day14.py
from collections import defaultdict
import math


def input():
    [template, rules] = open("input.txt", "r").read().split("\n\n")
    return (
        template,
        {
            key: value
            for [key, value] in [line.split(" -> ") for line in rules.splitlines()]
        },
    )


def run_insertions(pairs, rules):
    newPairs = pairs.copy()
    for key in pairs:
        if pairs[key] > 0 and rules[key]:
            newPairs[f"{key[0]}{rules[key]}"] += pairs[key]
            newPairs[f"{rules[key]}{key[1]}"] += pairs[key]
            newPairs[key] -= pairs[key]
    return newPairs


def do_the_puzzle(times):
    (template, rules) = input()

    pairs = defaultdict(lambda: 0)
    for pair in rules:
        pairs[pair] = 0
    for pair in [
        f"{template[i]}{template[i + 1]}" for i in range(0, len(template) - 1)
    ]:
        pairs[pair] += 1

    for _ in range(0, times):
        pairs = run_insertions(pairs, rules)

    occurances = defaultdict(lambda: 0)
    for pair in pairs:
        occurances[pair[0]] += pairs[pair]
    occurances[template[-1]] += 1

    occurances = [occurances[letter] for letter in occurances]
    return max(occurances) - min(occurances)


def part1():
    solution = do_the_puzzle(10)
    print("Part 1")
    print("   ", solution)
    print("   ", solution == 3143)


def part2():
    solution = do_the_puzzle(40)
    print("Part 2")
    print("   ", solution)


part1()
part2()
